tfhe/src/core_crypto/keycache.rs

```diff
@@ -0,0 +1,319 @@
+    }
+
+    pub fn clear_in_memory_cache(&self) {
+        self.inner.clear_in_memory_cache();
+    }
+}
+
+#[derive(Default)]
+pub struct KeyCache {
+    u32_multi_bit_cache: KeyCacheCoreImpl<MultiBitTestParams<u32>, MultiBitBootstrapKeys<u32>>,
+    u64_multi_bit_cache: KeyCacheCoreImpl<MultiBitTestParams<u64>, MultiBitBootstrapKeys<u64>>,
+    u32_classic_cache: KeyCacheCoreImpl<ClassicTestParams<u32>, ClassicBootstrapKeys<u32>>,
+    u64_classic_cache: KeyCacheCoreImpl<ClassicTestParams<u64>, ClassicBootstrapKeys<u64>>,
+    u128_classic_cache: KeyCacheCoreImpl<ClassicTestParams<u128>, ClassicBootstrapKeys<u64>>,
+    u32_fft_cache: KeyCacheCoreImpl<ClassicTestParams<u32>, FftBootstrapKeys<u32>>,
+    u64_fft_cache: KeyCacheCoreImpl<ClassicTestParams<u64>, FftBootstrapKeys<u64>>,
+    u128_fft_cache: KeyCacheCoreImpl<ClassicTestParams<u128>, FftBootstrapKeys<u128>>,
+    u64_fft_wopbs_cache: KeyCacheCoreImpl<FftWopPbsTestParams<u64>, FftWopPbsKeys<u64>>,
+    u32_pksk_cache: KeyCacheCoreImpl<PackingKeySwitchTestParams<u32>, PackingKeySwitchKeys<u32>>,
+    u64_pksk_cache: KeyCacheCoreImpl<PackingKeySwitchTestParams<u64>, PackingKeySwitchKeys<u64>>,
+}
+
+impl KeyCache {
+    pub fn get_key_with_closure<C, P, K>(&self, params: P, c: C) -> K
+    where
+        C: FnMut(P) -> K,
+        P: KeyCacheAccess<Keys = K> + Serialize + DeserializeOwned + Copy + PartialEq + NamedParam,
+        K: DeserializeOwned + Serialize + Clone,
+    {
+        P::access(self).get_key_with_closure(params, c)
+    }
+
+    pub fn clear_in_memory_cache<P, K>(&self)
+    where
+        P: KeyCacheAccess<Keys = K> + Serialize + DeserializeOwned + Copy + PartialEq + NamedParam,
+        K: DeserializeOwned + Serialize + Clone,
+    {
+        P::access(self).clear_in_memory_cache();
+    }
+}
+
+pub trait KeyCacheAccess: Serialize + DeserializeOwned + Copy + PartialEq + NamedParam {
+    type Keys: DeserializeOwned + Serialize;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys>;
+}
+
+impl<Scalar: UnsignedInteger + Serialize + DeserializeOwned> KeyCacheAccess
+    for MultiBitTestParams<Scalar>
+{
+    type Keys = MultiBitBootstrapKeys<Scalar>;
+
+    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
+        use std::any::TypeId;
+
+        let scalar_type_id = TypeId::of::<Scalar>();
+
+        if scalar_type_id == TypeId::of::<u32>() {
+            // unsafe { std::mem::transmute(&keycache.u32_multi_bit_cache) }
+            unsafe {
+                &*(std::ptr::addr_of!(keycache.u32_multi_bit_cache)
+                    as *const KeyCacheCoreImpl<Self, Self::Keys>)
+            }
+        } else if scalar_type_id == TypeId::of::<u64>() {
+            unsafe {
+                &*(std::ptr::addr_of!(keycache.u64_multi_bit_cache)
+                    as *const KeyCacheCoreImpl<Self, Self::Keys>)
+            }
+        } else {
+            panic!("No keycache for given Scalar type")
+        }
+    }
+}
```

Comment:
I think you could avoid use of unsafe by doing something like 
```rust
impl<Scalar: UnsignedInteger + Serialize + DeserializeOwned> KeyCacheAccess
    for MultiBitTestParams<u32>
{
    type Keys = MultiBitBootstrapKeys<u32>;

    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
        use std::any::TypeId;

    &keycache.u32_multi_bit_cache
  }
}

impl<Scalar: UnsignedInteger + Serialize + DeserializeOwned> KeyCacheAccess
    for MultiBitTestParams<u64>
{
    type Keys = MultiBitBootstrapKeys<u64>;

    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
        use std::any::TypeId;

    &keycache.u32_multi_bit_cache
  }
}
```

Its more code tho

Reply:
that would not work from the gen key cache thing though if the tests are generic over Scalar ? (is that the more code you mention ?)

Reply:
It will work
```rust

#[derive(Default)]
struct KeyCacheImpl<T>(T);

#[derive(Default)]
struct KeyCache{
    u32: KeyCacheImpl<u32>,
    u64: KeyCacheImpl<u64>,
}


trait KeyCacheAccess: Sized {
    fn access(kc: &KeyCache) -> &KeyCacheImpl<Self>;
}


impl KeyCache {
    fn get<Scalar>(&self) -> Scalar 
    where Scalar: KeyCacheAccess + Copy {
        Scalar::access(self).0
    }
}

impl KeyCacheAccess for u64 {
    fn access(kc: &KeyCache) -> &KeyCacheImpl<Self> {
        &kc.u64
    }
}

impl KeyCacheAccess for u32 {
    fn access(kc: &KeyCache) -> &KeyCacheImpl<Self> {
        &kc.u32
    }
}



fn test<Scalar: KeyCacheAccess + Copy + std::fmt::Display>(kc: &KeyCache) {
    println!("{}", kc.get::<Scalar>());
} 


fn main() {
    let keycache = KeyCache::default();
    test::<u32>(&keycache);
    test::<u64>(&keycache);
}
```

Reply:
So we would need to parameterize with the parameters, as we have several parameter sets for tests which are themselves generic over the scalar

Reply:
But yeah this has a chance to work well

Reply:
but I know rust would not let us do that at some point because the keycache was not implemented for the Scalar trait with a blanket implementation but rather for specific types implementing the trait

Reply:
That sounds extremely strange as  that not how traits should works

Reply:
we were not following the exact same pattern as you so to be checked

Reply:
it is as I remember, when instantiating the test with a set of parameter the key cache access implementation is not generic enough as it's not implemented for all scalar implementing types

```
error[E0277]: the trait bound `params::MultiBitTestParams<Scalar>: KeyCacheAccess` is not satisfied
   --> tfhe/src/core_crypto/algorithms/test/lwe_multi_bit_programmable_bootstrapping.rs:107:18
    |
107 |     let shared = gen_keys_or_get_from_cache_if_enabled(params, &mut keys_gen);
    |                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `KeyCacheAccess` is not implemented for `params::MultiBitTestParams<Scalar>`
```
if you have an implementation with 

```rust
impl KeyCacheAccess for MultiBitTestParams<u32> {
    type Keys = MultiBitBootstrapKeys<u32>;

    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
        &keycache.u32_multi_bit_cache
    }
}

impl KeyCacheAccess for MultiBitTestParams<u64> {
    type Keys = MultiBitBootstrapKeys<u64>;

    fn access(keycache: &KeyCache) -> &KeyCacheCoreImpl<Self, Self::Keys> {
        &keycache.u64_multi_bit_cache
    }
}
```

and test instantiation with concrete parameters

Reply:
example of test function being instantiated

```rust
fn std_lwe_encrypt_multi_bit_deterministic_pbs_decrypt_custom_mod<
    Scalar: UnsignedTorus + Sync + Send + CastFrom<usize> + CastInto<usize> + Serialize + DeserializeOwned,
>(
    params: MultiBitTestParams<Scalar>,
) {
```

Reply:
one possibility might be to return a Result to an &dyn Keycache, we return an error if such a keycache is not available (for u128 for example) and return a trait object otherwise maybe

Reply:
Don't worry, @tmontaigu walked me through his idea. I've implemented it and it works like a charm.

Reply:
I think I saw, and indeed this is very smart

Reply:
Problem is we need to have the keycache always enabled with this solution.
```rust
error[E0432]: unresolved import `crate::core_crypto::keycache`
  --> tfhe/src/core_crypto/fft_impl/common.rs:94:29
   |
94 |     use crate::core_crypto::keycache::KeyCacheAccess;
   |                             ^^^^^^^^ could not find `keycache` in `core_crypto`

error[E0405]: cannot find trait `KeyCacheAccess` in this scope
  --> tfhe/src/core_crypto/algorithms/test/lwe_multi_bit_programmable_bootstrapping.rs:78:33
   |
78 |     MultiBitTestParams<Scalar>: KeyCacheAccess<Keys = MultiBitBootstrapKeys<Scalar>>,
   |                                 ^^^^^^^^^^^^^^ not found in this scope


```

Reply:
so it's an any enable but as long as the internal-keycache feature is not given it's good right ?